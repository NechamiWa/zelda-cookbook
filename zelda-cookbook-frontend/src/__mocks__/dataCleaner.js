/* eslint-disable */
export const getIngredientNames = jest.fn()
  .mockImplementation(() => [{
    name: 'Apple',
    id: 1
  }])
