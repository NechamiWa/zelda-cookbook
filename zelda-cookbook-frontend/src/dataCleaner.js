export const getIngredientNames = rawData => {
  return rawData.map(({name, id}) => ({name, id}))
}
