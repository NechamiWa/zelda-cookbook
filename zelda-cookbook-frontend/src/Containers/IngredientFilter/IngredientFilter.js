import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { getIngredientNames } from '../../dataCleaner';
import IngredientTag from '../../Components/IngredientTag/IngredientTag';
import {
  addIngredientFilter,
  removeIngredientFilter
} from '../../actions';


export class IngredientFilter extends Component {
  constructor(props) {
    super(props);
    this.state = {
      ingredientSearch: '',
      searchFocus: false,
    };
  }

  handleChange = event => {
    const ingredientSearch = event.target.value;
    this.setState({ ingredientSearch });
  }

  handleFocus = () => {
    this.setState({ searchFocus: true });
  }

  handleBlur = () => {
    setTimeout(() => {
      this.setState({ searchFocus: false });
    }, 150);
  }

  ingredientListToRender = () => {
    const ingredientNames = getIngredientNames(this.props.ingredients);
    return ingredientNames
      .filter(food => food.name.toLowerCase().includes(
        this.state.ingredientSearch.toLowerCase())
      )
      .map(food => <p
        key={food.id} 
        className='search-results__item'
        id={food.id}
        onClick={this.selectIngredient}
      >
        {food.name}
      </p>);
  }

  selectIngredient = (event) => {
    const { ingredientFilter } = this.props;
    const { innerText , id } = event.target;
    const newIngredient = { id, name: innerText };
    const checkDupe = ingredientFilter.some(tag => tag.name === newIngredient.name);
    if ( !checkDupe && ingredientFilter.length < 5 ) {
      this.props.addIngredientFilter(newIngredient);
    }
    this.setState({ searchFocus: false, ingredientSearch: '' });
  }

  tagsToRender = () => {
    const { ingredientFilter } = this.props;
    return ingredientFilter.map(tag => <IngredientTag
      {...tag}
      key={tag.id}
      remove={this.removeIngredient}
    />);
  }

  removeIngredient = (event) => {
    const { id } = event.target;
    this.props.removeIngredientFilter(id);
  }

  render() {
    return (
      <div>
        
        <div 
          className='search-bar__ingredient'
          onFocus={this.handleFocus}
          onBlur={this.handleBlur}
        >
          <input
            type='text'
            placeholder='Select up to 5'
            value={this.state.ingredientSearch}
            className='search-bar'
            onChange={this.handleChange}
          />
          <section className={`
            search-results__container
            ${this.state.searchFocus ? 'search-results__containter--show' : ''}
          `}>
            {this.ingredientListToRender()}
          </section>
        </div>
        <section className='ingredientTag-container'>
          {this.tagsToRender()}
        </section>
      </div>
    );
  }
}

const { arrayOf, shape, string, number, array, func } = PropTypes;

IngredientFilter.propTypes = {
  ingredients: arrayOf(shape({
    name: string,
    id: number
  })),
  ingredientFilter: array,
  addIngredientFilter: func,
  removeIngredientFilter: func,
};

export const MSTP = ({ingredients, ingredientFilter}) => ({
  ingredients,
  ingredientFilter
});

export const MDTP = dispatch => ({
  addIngredientFilter: ingredient => dispatch(addIngredientFilter(ingredient)),
  removeIngredientFilter: id => dispatch(removeIngredientFilter(id))
});

export default connect(MSTP, MDTP)(IngredientFilter);