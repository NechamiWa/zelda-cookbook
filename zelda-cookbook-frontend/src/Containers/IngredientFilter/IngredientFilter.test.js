/* eslint-disable */
import React from 'react';
import { shallow } from 'enzyme';
import { IngredientFilter, MSTP, MDTP } from './IngredientFilter';
import * as dataCleaner from '../../dataCleaner';
import {
  addIngredientFilter,
  removeIngredientFilter,
} from '../../actions';

jest.mock('../../dataCleaner');
jest.useFakeTimers();


describe('IngredientFilter', () => {
  let wrapper;
  let mockIngredients = [];
  let mockIngredientFilter = [];
  let mockAddIngredientFilter = jest.fn();
  let mockRemoveIngredientFilter = jest.fn();

  beforeEach(() => {
    wrapper = shallow(<IngredientFilter
      ingredients={ mockIngredients }
      ingredientFilter = { mockIngredientFilter }
      addIngredientFilter = { mockAddIngredientFilter }
      removeIngredientFilter = { mockRemoveIngredientFilter }
    />);
  });

  it('should have a default state', () => {
    const expected = {
      ingredientSearch: '',
      searchFocus: false
    };

    expect(wrapper.state()).toEqual(expected);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('handleChange', () => {
    it('should set ingredientSearch in state', () => {
      const mockEvent = { target: { 
        value: 'qwerty'
      }};

      wrapper.instance().handleChange(mockEvent);
      expect(wrapper.state().ingredientSearch).toEqual('qwerty');
    });
  });

  describe('handleFocus', () => {
    it('should set searchFocus to true', () => {
      wrapper.instance().handleFocus()
      expect(wrapper.state().searchFocus).toEqual(true)
    });
  });

  describe('handleBlur', () => {
    it('should set searchFocus to false', () => {
      wrapper.state().searchFocus = true
      wrapper.instance().handleBlur()
      jest.runAllTimers()
      expect(wrapper.state().searchFocus).toEqual(false)
    });
  });

  describe('ingredientListToRender', () => {
    it('should render a list of ingredients and match the snapshot', () => {
      expect(wrapper.instance().ingredientListToRender()).toMatchSnapshot()
    });

    it('should filter its results based on what the user is searching for', () => {
      wrapper.state().ingredientSearch = 'asdf'
      expect(wrapper.instance().ingredientListToRender()).toMatchSnapshot()
    });
  });

  describe('selectIngredient', () => {
    it('should not call addIngredientFilter if the object is already selected', () => {
      wrapper = shallow(<IngredientFilter
        ingredients={ mockIngredients }
        ingredientFilter = { [{id: '1', name: 'Apple'}] }
        addIngredientFilter = { mockAddIngredientFilter }
        removeIngredientFilter = { mockRemoveIngredientFilter }
      />)

      const mockEvent = { target: {id: '1', innerText: 'Apple'}}
      wrapper.instance().selectIngredient(mockEvent)
      expect(mockAddIngredientFilter).not.toHaveBeenCalled()
    });

    it('should not call addIngredientFilter if there are 5 ingredients already', () => {
      wrapper = shallow(<IngredientFilter
        ingredients={ mockIngredients }
        ingredientFilter = {[
          {id: '0', name: 'Chicken'},
          {id: '2', name: 'Grape'},
          {id: '3', name: 'Pear'},
          {id: '4', name: 'Fish'},
          {id: '5', name: 'Tuna'}
        ]}
        addIngredientFilter = { mockAddIngredientFilter }
        removeIngredientFilter = { mockRemoveIngredientFilter }
      />)
      const mockEvent = { 
        target: {
          id: '1', innerText: 'Apple'
        }
      }
      wrapper.instance().selectIngredient(mockEvent)
      expect(mockAddIngredientFilter).not.toHaveBeenCalled()
    });

    it('should call this.props.addIngredientFilter with a new ingredient', () => {
      const mockEvent = { target: {id: '1', innerText: 'Apple'}}
      const expected = {id: '1', name: 'Apple'}
      wrapper.instance().selectIngredient(mockEvent)
      expect(mockAddIngredientFilter).toHaveBeenCalledWith(expected)
    });
  });

  describe('tagsToRender', () => {
    it('should return tags depending on what is in the filters ingredients array', () => {
      wrapper = shallow(<IngredientFilter
        ingredients={ mockIngredients }
        ingredientFilter = {[
          {id: '0', name: 'Chicken'},
          {id: '2', name: 'Grape'},
          {id: '3', name: 'Pear'},
          {id: '4', name: 'Fish'},
          {id: '5', name: 'Tuna'}
        ]}
        addIngredientFilter = { mockAddIngredientFilter }
        removeIngredientFilter = { mockRemoveIngredientFilter }
      />)
      expect(wrapper.instance().tagsToRender()).toMatchSnapshot()
    });
  });

  describe('removeIngredient', () => {
    it('should call removeIngredientFilter with the selected id', () => {
      const mockEvent = {target: {id: 1}}

      wrapper.instance().removeIngredient(mockEvent)
      expect(mockRemoveIngredientFilter).toHaveBeenCalledWith(1)
    });
  });
});

describe('MSTP', () => {
  it('should return an object with ingredients and a filter object', () => {
    const mockState = {
      ingredients: [{
        id: '1',
        name: 'Apple',
      }],
      ingredientFilter: []
    }
    const mapped = MSTP(mockState);

    expect(mapped.ingredients).toEqual([{
      id: '1',
      name: 'Apple',
    }]);
    expect(mapped.ingredientFilter).toEqual([]);
  });
});

describe('MDTP', () => {
  it('should call dispatch when addIngredientFilter or removeIngredientFilter are called', () => {
    const mockDispatch = jest.fn();
    const mockIngredient = { id: '1', name: 'Apple' };
    const mockId = 1;
    const result = MDTP(mockDispatch);

    result.addIngredientFilter(mockIngredient);
    expect(mockDispatch).toHaveBeenCalledWith(addIngredientFilter(mockIngredient));

    result.removeIngredientFilter(mockId);
    expect(mockDispatch).toHaveBeenCalledWith(removeIngredientFilter(mockId));
  });
});