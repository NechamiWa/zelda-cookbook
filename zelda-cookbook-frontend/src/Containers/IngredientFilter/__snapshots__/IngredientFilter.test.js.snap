// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IngredientFilter ingredientListToRender should filter its results based on what the user is searching for 1`] = `Array []`;

exports[`IngredientFilter ingredientListToRender should render a list of ingredients and match the snapshot 1`] = `
Array [
  <p
    className="search-results__item"
    id={1}
    onClick={[Function]}
>
    Apple
</p>,
]
`;

exports[`IngredientFilter should match the snapshot 1`] = `
<div>
  <div
    className="search-bar__ingredient"
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <input
      className="search-bar"
      onChange={[Function]}
      placeholder="Select up to 5"
      type="text"
      value=""
    />
    <section
      className="
                search-results__container
                
              "
    >
      <p
        className="search-results__item"
        id={1}
        key="1"
        onClick={[Function]}
      >
        Apple
      </p>
    </section>
  </div>
  <section
    className="ingredientTag-container"
  />
</div>
`;

exports[`IngredientFilter tagsToRender should return tags depending on what is in the filters ingredients array 1`] = `
Array [
  <IngredientTag
    id="0"
    name="Chicken"
    remove={[Function]}
/>,
  <IngredientTag
    id="2"
    name="Grape"
    remove={[Function]}
/>,
  <IngredientTag
    id="3"
    name="Pear"
    remove={[Function]}
/>,
  <IngredientTag
    id="4"
    name="Fish"
    remove={[Function]}
/>,
  <IngredientTag
    id="5"
    name="Tuna"
    remove={[Function]}
/>,
]
`;
